pkg/controller/selector/filter.go

```diff
@@ -15,191 +15,304 @@
 package selector
 
 import (
+	"encoding/json"
 	"errors"
+	"regexp"
+	"strconv"
+	"strings"
 
+	log "github.com/golang/glog"
 	"github.com/opensds/opensds/pkg/model"
 )
 
-type Filter interface {
-	Handle(request map[string]interface{}, pools []*model.StoragePoolSpec) ([]*model.StoragePoolSpec, error)
+// StructToMap ...
+func StructToMap(structObj interface{}) (map[string]interface{}, error) {
+	jsonStr, err := json.Marshal(structObj)
+	if nil != err {
+		return nil, err
+	}
+
+	var result map[string]interface{}
+	err = json.Unmarshal(jsonStr, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	for key, value := range result {
+		valueMap, ok := value.(map[string]interface{})
+		if ok {
+			for k, v := range valueMap {
+				result[key+"."+k] = v
+			}
+			delete(result, key)
+		}
+	}
+
+	return result, nil
 }
 
-type AZFilter struct {
-	Next Filter
+// Epsilon ...
+const Epsilon float64 = 0.00000001
+
+// IsFloatEqual ...
+func IsFloatEqual(a, b float64) bool {
+	if (a-b) < Epsilon && (b-a) < Epsilon {
+		return true
+	}
+
+	return false
 }
 
-func (filter *AZFilter) Handle(request map[string]interface{}, pools []*model.StoragePoolSpec) ([]*model.StoragePoolSpec, error) {
-	availabilityZone, ok := request["availabilityZone"]
-	if !ok {
-		return nil, errors.New("The request doesn't contain availability zone property when filter pools in AZFilter.")
+// IsAvailablePool ...
+func IsAvailablePool(filterReq map[string]interface{}, pool *model.StoragePoolSpec) (bool, error) {
+	poolMap, err := StructToMap(pool)
+	if nil != err {
+		return false, err
 	}
 
-	availablePools := []*model.StoragePoolSpec{}
-	for _, pool := range pools {
-		if availabilityZone == pool.AvailabilityZone {
-			availablePools = append(availablePools, pool)
+	for key, reqValue := range filterReq {
+		poolValue, ok := poolMap[key]
+		if !ok {
+			log.Info("pool: " + pool.Name + " doesn't provide capability: " + key)
+			return false, nil
 		}
-	}
 
-	if len(availablePools) == 0 {
-		return nil, errors.New("No available pools to meet user's availability zone requirement.")
-	} else if filter.Next != nil {
-		return filter.Next.Handle(request, availablePools)
+		ismatch, err := match(key, poolValue, reqValue)
+		if nil != err {
+			return false, err
+		}
+
+		if false == ismatch {
+			return false, nil
+		}
 	}
-	return availablePools, nil
-}
 
-type CapacityFilter struct {
-	Next Filter
+	return true, nil
 }
 
-func (filter *CapacityFilter) Handle(request map[string]interface{}, pools []*model.StoragePoolSpec) ([]*model.StoragePoolSpec, error) {
-
-	size, ok := request["size"]
+// match ...
+func match(key string, value interface{}, reqValue interface{}) (bool, error) {
```

Comment:
I think we should have a detailed discussion about the filter rule, since it's related to capability report ([#281](https://github.com/opensds/opensds/pull/281)) feature.

Reply:
I believe this design follows Cinder's scheduler. So it is actually not necessary to specify all the parameters in the capability reporting in the scheduler. Matching should be done automatically between the profile and the capabilities reported by the backend. That being said, we do need a design doc that has detailed explanations on how this work.

Reply:
The design document is being written.